File: src/narrative/evaluators/pyramid_evaluator.py
# -*- coding: utf-8 -*-
"""
金字塔原理评估器 (Pyramid Evaluator)

评估文本的逻辑结构，基于芭芭拉·明托的金字塔原理：
1. 结论先行
2. 纵向结构（问答关系）
3. 横向结构（逻辑排列）
4. MECE原则
"""

from typing import Dict, Any, Optional, List
from .base import BaseEvaluator, EvaluationResult, Issue, Severity, ScoreLevel


class PyramidEvaluator(BaseEvaluator):
    """
    金字塔原理评估器
    
    评估维度：
    - conclusion_first: 结论先行程度
    - vertical_structure: 纵向逻辑（问答关系）
    - horizontal_structure: 横向逻辑（排列顺序）
    - mece_compliance: MECE原则遵循程度
    """
    
    @property
    def name(self) -> str:
        return "pyramid_evaluator"
    
    @property
    def description(self) -> str:
        return "评估文本的逻辑结构，检查金字塔原理的四大法则"
    
    @property
    def related_skill(self) -> str:
        return "pyramid-structure"
    
    # 结论先行的信号词
    CONCLUSION_SIGNALS = [
        "因此", "所以", "综上所述", "总之", "结论是", "我们认为",
        "建议", "应该", "必须", "核心观点是", "关键是", "重点是",
        "本文主张", "我们的结论", "答案是", "解决方案是"
    ]
    
    # 埋没结论的信号词（出现在开头是警告）
    BURIED_CONCLUSION_SIGNALS = [
        "首先让我们看看", "在讨论之前", "要理解这个问题",
        "背景是", "历史上", "众所周知", "一般来说"
    ]
    
    # 逻辑连接词
    LOGICAL_CONNECTORS = {
        "递进": ["此外", "而且", "更重要的是", "进一步", "不仅如此"],
        "因果": ["因为", "由于", "所以", "因此", "导致", "使得"],
        "转折": ["但是", "然而", "不过", "尽管如此", "相反"],
        "并列": ["同时", "另外", "一方面...另一方面", "既...又"],
        "总结": ["总之", "综上", "概括来说", "简言之"]
    }
    
    # 结构标记词
    STRUCTURE_MARKERS = [
        "第一", "第二", "第三", "首先", "其次", "最后",
        "一是", "二是", "三是", "1.", "2.", "3.",
        "一、", "二、", "三、"
    ]
    
    async def evaluate(
        self,
        content: str,
        context: Optional[Dict[str, Any]] = None
    ) -> EvaluationResult:
        """执行金字塔原理评估"""
        
        issues = []
        metrics = {}
        
        # 分段分析
        paragraphs = [p.strip() for p in content.split('\n\n') if p.strip()]
        sentences = self._split_sentences(content)
        
        # 1. 评估结论先行
        conclusion_score, conclusion_issues = self._evaluate_conclusion_first(
            content, paragraphs, sentences
        )
        metrics["conclusion_first"] = conclusion_score
        issues.extend(conclusion_issues)
        
        # 2. 评估纵向结构
        vertical_score, vertical_issues = self._evaluate_vertical_structure(
            content, paragraphs
        )
        metrics["vertical_structure"] = vertical_score
        issues.extend(vertical_issues)
        
        # 3. 评估横向结构
        horizontal_score, horizontal_issues = self._evaluate_horizontal_structure(
            content, paragraphs
        )
        metrics["horizontal_structure"] = horizontal_score
        issues.extend(horizontal_issues)
        
        # 4. 评估MECE原则
        mece_score, mece_issues = self._evaluate_mece(content, paragraphs)
        metrics["mece_compliance"] = mece_score
        issues.extend(mece_issues)
        
        # 计算总分（加权平均）
        weights = {
            "conclusion_first": 0.30,      # 结论先行最重要
            "vertical_structure": 0.25,
            "horizontal_structure": 0.25,
            "mece_compliance": 0.20
        }
        
        total_score = sum(
            metrics[k] * weights[k] for k in weights
        )
        
        # 生成摘要
        summary = self._generate_summary(total_score, metrics, issues)
        
        return EvaluationResult(
            evaluator_name=self.name,
            score=total_score,
            level=self._score_to_level(total_score),
            issues=issues,
            metrics=metrics,
            summary=summary
        )
    
    def _evaluate_conclusion_first(
        self,
        content: str,
        paragraphs: List[str],
        sentences: List[str]
    ) -> tuple:
        """评估结论先行程度"""
        issues = []
        score = 50  # 基础分
        
        if not paragraphs or not sentences:
            return score, issues
        
        first_para = paragraphs[0]
        first_sentences = sentences[:3] if len(sentences) >= 3 else sentences
        first_text = ' '.join(first_sentences)
        
        # 检查开头是否有结论信号
        has_conclusion_signal = any(
            signal in first_text for signal in self.CONCLUSION_SIGNALS
        )
        
        if has_conclusion_signal:
            score += 30
        else:
            issues.append(Issue(
                code="PYRAMID_NO_CONCLUSION_FIRST",
                message="开篇未明确陈述核心结论，读者可能需要阅读全文才能理解重点",
                severity=Severity.MAJOR,
                location="开头段落",
                suggestion="在文章开头明确陈述你的核心观点或结论",
                related_skill=self.related_skill
            ))
        
        # 检查是否埋没结论
        has_buried_signal = any(
            signal in first_para[:100] for signal in self.BURIED_CONCLUSION_SIGNALS
        )
        
        if has_buried_signal:
            score -= 15
            issues.append(Issue(
                code="PYRAMID_BURIED_CONCLUSION",
                message="开篇使用了背景铺垫式写法，结论可能被埋没",
                severity=Severity.MINOR,
                location="开头段落",
                suggestion="考虑将结论前置，背景信息可以作为支撑放在后面",
                related_skill=self.related_skill
            ))
        
        # 检查是否有结构预告
        structure_preview = any(
            marker in first_para for marker in ["三个", "几个方面", "以下", "将从"]
        )
        
        if structure_preview:
            score += 20
        
        return min(100, max(0, score)), issues
    
    def _evaluate_vertical_structure(
        self,
        content: str,
        paragraphs: List[str]
    ) -> tuple:
        """评估纵向结构（问答关系）"""
        issues = []
        score = 60  # 基础分
        
        # 检查因果连接词的使用
        causal_connectors = self.LOGICAL_CONNECTORS["因果"]
        causal_count = sum(
            content.count(conn) for conn in causal_connectors
        )
        
        if causal_count >= 3:
            score += 20
        elif causal_count >= 1:
            score += 10
        else:
            issues.append(Issue(
                code="PYRAMID_WEAK_CAUSAL_CHAIN",
                message="论证缺乏明确的因果链条，论点之间的逻辑关系不够清晰",
                severity=Severity.MINOR,
                location="全文",
                suggestion="使用'因为...所以...'等连接词明确论点间的逻辑关系",
                related_skill=self.related_skill
            ))
        
        # 检查是否有"为什么"的回应
        why_responses = ["原因是", "这是因为", "之所以", "理由是"]
        has_why_response = any(resp in content for resp in why_responses)
        
        if has_why_response:
            score += 15
        
        # 检查递进关系
        progressive_connectors = self.LOGICAL_CONNECTORS["递进"]
        progressive_count = sum(
            content.count(conn) for conn in progressive_connectors
        )
        
        if progressive_count >= 2:
            score += 5
        
        return min(100, max(0, score)), issues
    
    def _evaluate_horizontal_structure(
        self,
        content: str,
        paragraphs: List[str]
    ) -> tuple:
        """评估横向结构（逻辑排列）"""
        issues = []
        score = 60  # 基础分
        
        # 检查结构标记的使用
        marker_count = sum(
            1 for marker in self.STRUCTURE_MARKERS if marker in content
        )
        
        if marker_count >= 3:
            score += 25
        elif marker_count >= 2:
            score += 15
        elif marker_count == 0:
            issues.append(Issue(
                code="PYRAMID_NO_STRUCTURE_MARKERS",
                message="缺乏清晰的结构标记，同级论点的排列顺序不明确",
                severity=Severity.MINOR,
                location="全文",
                suggestion="使用'第一、第二、第三'或'首先、其次、最后'等标记组织论点",
                related_skill=self.related_skill
            ))
        
        # 检查并列结构
        parallel_markers = ["一方面", "另一方面", "同时", "此外"]
        parallel_count = sum(
            content.count(marker) for marker in parallel_markers
        )
        
        if parallel_count >= 2:
            score += 10
        
        # 检查总结句
        summary_markers = self.LOGICAL_CONNECTORS["总结"]
        has_summary = any(marker in content for marker in summary_markers)
        
        if has_summary:
            score += 5
        
        return min(100, max(0, score)), issues
    
    def _evaluate_mece(
        self,
        content: str,
        paragraphs: List[str]
    ) -> tuple:
        """评估MECE原则（相互独立，完全穷尽）"""
        issues = []
        score = 70  # 基础分，MECE难以自动精确评估
        
        # 检查是否有明确的分类框架
        framework_signals = [
            "从...角度", "分为...类", "包括以下几个方面",
            "可以分为", "主要有", "三个维度", "四个层面"
        ]
        
        has_framework = any(signal in content for signal in framework_signals)
        
        if has_framework:
            score += 15
        else:
            issues.append(Issue(
                code="PYRAMID_NO_CLEAR_FRAMEWORK",
                message="未使用明确的分类框架，可能存在论点重叠或遗漏",
                severity=Severity.INFO,
                location="全文",
                suggestion="考虑使用MECE框架（如3C、4P、What/Why/How）组织论点",
                related_skill=self.related_skill
            ))
        
        # 检查是否有重复词汇密集出现（可能暗示内容重叠）
        # 这是一个启发式检查，不完全准确
        words = content.replace('\n', ' ').split()
        if len(words) > 50:
            word_freq = {}
            for word in words:
                if len(word) > 2:  # 忽略短词
                    word_freq[word] = word_freq.get(word, 0) + 1
            
            # 如果某个词出现频率过高，可能暗示重复论述
            high_freq_words = [w for w, f in word_freq.items() if f > len(words) * 0.05]
            if len(high_freq_words) > 3:
                score -= 10
                issues.append(Issue(
                    code="PYRAMID_POSSIBLE_OVERLAP",
                    message="部分概念重复出现频率较高，可能存在论点重叠",
                    severity=Severity.INFO,
                    location="全文",
                    suggestion="检查各论点是否真正独立，避免重复论述相同内容",
                    related_skill=self.related_skill
                ))
        
        return min(100, max(0, score)), issues
    
    def _split_sentences(self, text: str) -> List[str]:
        """分句"""
        import re
        sentences = re.split(r'[。！？.!?]', text)
        return [s.strip() for s in sentences if s.strip()]
    
    def _generate_summary(
        self,
        score: float,
        metrics: Dict[str, float],
        issues: List[Issue]
    ) -> str:
        """生成评估摘要"""
        # 找出最弱维度
        weakest = min(metrics, key=metrics.get)
        strongest = max(metrics, key=metrics.get)
        
        dimension_names = {
            "conclusion_first": "结论先行",
            "vertical_structure": "纵向逻辑",
            "horizontal_structure": "横向结构",
            "mece_compliance": "MECE原则"
        }
        
        summary = f"逻辑结构得分：{score:.1f}/100。"
        summary += f"最强：{dimension_names.get(strongest, strongest)}，"
        summary += f"最弱：{dimension_names.get(weakest, weakest)}。"
        
        critical_count = sum(1 for i in issues if i.severity == Severity.CRITICAL)
        major_count = sum(1 for i in issues if i.severity == Severity.MAJOR)
        
        if critical_count > 0:
            summary += f"发现{critical_count}个严重结构问题需要修复。"
        elif major_count > 0:
            summary += f"发现{major_count}个主要结构问题建议优化。"
        
        return summary
    
    def quick_scan(self, content: str) -> EvaluationResult:
        """快速扫描（不使用LLM）"""
        issues = []
        
        # 简单检查：开头是否有结论信号
        first_200 = content[:200] if len(content) > 200 else content
        has_conclusion = any(
            signal in first_200 for signal in self.CONCLUSION_SIGNALS
        )
        
        # 检查结构标记
        has_structure = any(
            marker in content for marker in self.STRUCTURE_MARKERS
        )
        
        score = 50
        if has_conclusion:
            score += 25
        else:
            issues.append(Issue(
                code="PYRAMID_NO_CONCLUSION_FIRST",
                message="开篇可能未明确陈述结论",
                severity=Severity.MAJOR,
                related_skill=self.related_skill
            ))
        
        if has_structure:
            score += 25
        else:
            issues.append(Issue(
                code="PYRAMID_NO_STRUCTURE_MARKERS",
                message="缺乏结构标记",
                severity=Severity.MINOR,
                related_skill=self.related_skill
            ))
        
        return EvaluationResult(
            evaluator_name=self.name,
            score=score,
            level=self._score_to_level(score),
            issues=issues,
            summary=f"快速扫描：逻辑结构得分{score}/100"
        )

File: src-ts/narrative/evaluators/pyramid-evaluator.ts
/**
 * Pyramid Evaluator
 *
 * Evaluates logical structure based on Barbara Minto's Pyramid Principle:
 * 1. Conclusion first
 * 2. Vertical structure (question-answer relationship)
 * 3. Horizontal structure (logical arrangement)
 * 4. MECE principle
 */

import {
  BaseEvaluator,
  EvaluationResult,
  Issue,
  Severity,
  ScoreLevel,
} from './base';

export class PyramidEvaluator extends BaseEvaluator {
  get name(): string {
    return 'pyramid_evaluator';
  }

  get description(): string {
    return 'Evaluates logical structure against the four rules of the Pyramid Principle';
  }

  get relatedSkill(): string {
    return 'pyramid-structure';
  }

  private static readonly CONCLUSION_SIGNALS: readonly string[] = [
    '\u56e0\u6b64',
    '\u6240\u4ee5',
    '\u7efc\u4e0a\u6240\u8ff0',
    '\u603b\u4e4b',
    '\u7ed3\u8bba\u662f',
    '\u6211\u4eec\u8ba4\u4e3a',
    '\u5efa\u8bae',
    '\u5e94\u8be5',
    '\u5fc5\u987b',
    '\u6838\u5fc3\u89c2\u70b9\u662f',
    '\u5173\u952e\u662f',
    '\u91cd\u70b9\u662f',
    '\u672c\u6587\u4e3b\u5f20',
    '\u6211\u4eec\u7684\u7ed3\u8bba',
    '\u7b54\u6848\u662f',
    '\u89e3\u51b3\u65b9\u6848\u662f',
  ];

  private static readonly BURIED_CONCLUSION_SIGNALS: readonly string[] = [
    '\u9996\u5148\u8ba9\u6211\u4eec\u770b\u770b',
    '\u5728\u8ba8\u8bba\u4e4b\u524d',
    '\u8981\u7406\u89e3\u8fd9\u4e2a\u95ee\u9898',
    '\u80cc\u666f\u662f',
    '\u5386\u53f2\u4e0a',
    '\u4f17\u6240\u5468\u77e5',
    '\u4e00\u822c\u6765\u8bf4',
  ];

  private static readonly LOGICAL_CONNECTORS: Readonly<
    Record<string, readonly string[]>
  > = {
    progressive: [
      '\u6b64\u5916',
      '\u800c\u4e14',
      '\u66f4\u91cd\u8981\u7684\u662f',
      '\u8fdb\u4e00\u6b65',
      '\u4e0d\u4ec5\u5982\u6b64',
    ],
    causal: [
      '\u56e0\u4e3a',
      '\u7531\u4e8e',
      '\u6240\u4ee5',
      '\u56e0\u6b64',
      '\u5bfc\u81f4',
      '\u4f7f\u5f97',
    ],
    transition: [
      '\u4f46\u662f',
      '\u7136\u800c',
      '\u4e0d\u8fc7',
      '\u5c3d\u7ba1\u5982\u6b64',
      '\u76f8\u53cd',
    ],
    parallel: [
      '\u540c\u65f6',
      '\u53e6\u5916',
      '\u4e00\u65b9\u9762...\u53e6\u4e00\u65b9\u9762',
      '\u65e2...\u53c8',
    ],
    summary: [
      '\u603b\u4e4b',
      '\u7efc\u4e0a',
      '\u6982\u62ec\u6765\u8bf4',
      '\u7b80\u8a00\u4e4b',
    ],
  };

  private static readonly STRUCTURE_MARKERS: readonly string[] = [
    '\u7b2c\u4e00',
    '\u7b2c\u4e8c',
    '\u7b2c\u4e09',
    '\u9996\u5148',
    '\u5176\u6b21',
    '\u6700\u540e',
    '\u4e00\u662f',
    '\u4e8c\u662f',
    '\u4e09\u662f',
    '1.',
    '2.',
    '3.',
    '\u4e00\u3001',
    '\u4e8c\u3001',
    '\u4e09\u3001',
  ];

  async evaluate(
    content: string,
    context?: Record<string, unknown>,
  ): Promise<EvaluationResult> {
    const issues: Issue[] = [];
    const metrics: Record<string, number> = {};

    const paragraphs = content
      .split('\n\n')
      .map((p) => p.trim())
      .filter((p) => p.length > 0);
    const sentences = this.splitSentences(content);

    const [conclusionScore, conclusionIssues] =
      this.evaluateConclusionFirst(content, paragraphs, sentences);
    metrics['conclusion_first'] = conclusionScore;
    issues.push(...conclusionIssues);

    const [verticalScore, verticalIssues] = this.evaluateVerticalStructure(
      content,
      paragraphs,
    );
    metrics['vertical_structure'] = verticalScore;
    issues.push(...verticalIssues);

    const [horizontalScore, horizontalIssues] =
      this.evaluateHorizontalStructure(content, paragraphs);
    metrics['horizontal_structure'] = horizontalScore;
    issues.push(...horizontalIssues);

    const [meceScore, meceIssues] = this.evaluateMece(content, paragraphs);
    metrics['mece_compliance'] = meceScore;
    issues.push(...meceIssues);

    const weights: Record<string, number> = {
      conclusion_first: 0.3,
      vertical_structure: 0.25,
      horizontal_structure: 0.25,
      mece_compliance: 0.2,
    };

    const totalScore = Object.entries(weights).reduce(
      (sum, [k, w]) => sum + (metrics[k] ?? 0) * w,
      0,
    );

    const summary = this.generateSummary(totalScore, metrics, issues);

    return new EvaluationResult(
      this.name,
      totalScore,
      this.scoreToLevel(totalScore),
      issues,
      metrics,
      summary,
    );
  }

  private evaluateConclusionFirst(
    content: string,
    paragraphs: string[],
    sentences: string[],
  ): [number, Issue[]] {
    const issues: Issue[] = [];
    let score = 50;

    if (paragraphs.length === 0 || sentences.length === 0) {
      return [score, issues];
    }

    const firstSentences = sentences.slice(0, Math.min(3, sentences.length));
    const firstText = firstSentences.join(' ');

    const hasConclusionSignal = PyramidEvaluator.CONCLUSION_SIGNALS.some(
      (s) => firstText.includes(s),
    );

    if (hasConclusionSignal) {
      score += 30;
    } else {
      issues.push({
        code: 'PYRAMID_NO_CONCLUSION_FIRST',
        message:
          '\u5f00\u7bc7\u672a\u660e\u786e\u9648\u8ff0\u6838\u5fc3\u7ed3\u8bba\uff0c\u8bfb\u8005\u53ef\u80fd\u9700\u8981\u9605\u8bfb\u5168\u6587\u624d\u80fd\u7406\u89e3\u91cd\u70b9',
        severity: Severity.MAJOR,
        location: '\u5f00\u5934\u6bb5\u843d',
        suggestion:
          '\u5728\u6587\u7ae0\u5f00\u5934\u660e\u786e\u9648\u8ff0\u4f60\u7684\u6838\u5fc3\u89c2\u70b9\u6216\u7ed3\u8bba',
        relatedSkill: this.relatedSkill,
      });
    }

    const firstPara = paragraphs[0];
    const hasBuriedSignal = PyramidEvaluator.BURIED_CONCLUSION_SIGNALS.some(
      (s) => firstPara.slice(0, 100).includes(s),
    );

    if (hasBuriedSignal) {
      score -= 15;
      issues.push({
        code: 'PYRAMID_BURIED_CONCLUSION',
        message:
          '\u5f00\u7bc7\u4f7f\u7528\u4e86\u80cc\u666f\u94fa\u57ab\u5f0f\u5199\u6cd5\uff0c\u7ed3\u8bba\u53ef\u80fd\u88ab\u57cb\u6ca1',
        severity: Severity.MINOR,
        location: '\u5f00\u5934\u6bb5\u843d',
        suggestion:
          '\u8003\u8651\u5c06\u7ed3\u8bba\u524d\u7f6e\uff0c\u80cc\u666f\u4fe1\u606f\u53ef\u4ee5\u4f5c\u4e3a\u652f\u6491\u653e\u5728\u540e\u9762',
        relatedSkill: this.relatedSkill,
      });
    }

    const structurePreviewWords = [
      '\u4e09\u4e2a',
      '\u51e0\u4e2a\u65b9\u9762',
      '\u4ee5\u4e0b',
      '\u5c06\u4ece',
    ];
    if (structurePreviewWords.some((w) => firstPara.includes(w))) {
      score += 20;
    }

    return [Math.min(100, Math.max(0, score)), issues];
  }

  private evaluateVerticalStructure(
    content: string,
    paragraphs: string[],
  ): [number, Issue[]] {
    const issues: Issue[] = [];
    let score = 60;

    const causalConnectors = PyramidEvaluator.LOGICAL_CONNECTORS['causal'];
    const causalCount = causalConnectors.reduce(
      (sum, conn) => sum + content.split(conn).length - 1,
      0,
    );

    if (causalCount >= 3) {
      score += 20;
    } else if (causalCount >= 1) {
      score += 10;
    } else {
      issues.push({
        code: 'PYRAMID_WEAK_CAUSAL_CHAIN',
        message:
          '\u8bba\u8bc1\u7f3a\u4e4f\u660e\u786e\u7684\u56e0\u679c\u94fe\u6761\uff0c\u8bba\u70b9\u4e4b\u95f4\u7684\u903b\u8f91\u5173\u7cfb\u4e0d\u591f\u6e05\u6670',
        severity: Severity.MINOR,
        location: '\u5168\u6587',
        suggestion:
          "\u4f7f\u7528'\u56e0\u4e3a...\u6240\u4ee5...'\u7b49\u8fde\u63a5\u8bcd\u660e\u786e\u8bba\u70b9\u95f4\u7684\u903b\u8f91\u5173\u7cfb",
        relatedSkill: this.relatedSkill,
      });
    }

    const whyResponses = [
      '\u539f\u56e0\u662f',
      '\u8fd9\u662f\u56e0\u4e3a',
      '\u4e4b\u6240\u4ee5',
      '\u7406\u7531\u662f',
    ];
    if (whyResponses.some((r) => content.includes(r))) {
      score += 15;
    }

    const progressiveConnectors =
      PyramidEvaluator.LOGICAL_CONNECTORS['progressive'];
    const progressiveCount = progressiveConnectors.reduce(
      (sum, conn) => sum + content.split(conn).length - 1,
      0,
    );

    if (progressiveCount >= 2) {
      score += 5;
    }

    return [Math.min(100, Math.max(0, score)), issues];
  }

  private evaluateHorizontalStructure(
    content: string,
    paragraphs: string[],
  ): [number, Issue[]] {
    const issues: Issue[] = [];
    let score = 60;

    const markerCount = PyramidEvaluator.STRUCTURE_MARKERS.filter((m) =>
      content.includes(m),
    ).length;

    if (markerCount >= 3) {
      score += 25;
    } else if (markerCount >= 2) {
      score += 15;
    } else if (markerCount === 0) {
      issues.push({
        code: 'PYRAMID_NO_STRUCTURE_MARKERS',
        message:
          '\u7f3a\u4e4f\u6e05\u6670\u7684\u7ed3\u6784\u6807\u8bb0\uff0c\u540c\u7ea7\u8bba\u70b9\u7684\u6392\u5217\u987a\u5e8f\u4e0d\u660e\u786e',
        severity: Severity.MINOR,
        location: '\u5168\u6587',
        suggestion:
          "\u4f7f\u7528'\u7b2c\u4e00\u3001\u7b2c\u4e8c\u3001\u7b2c\u4e09'\u6216'\u9996\u5148\u3001\u5176\u6b21\u3001\u6700\u540e'\u7b49\u6807\u8bb0\u7ec4\u7ec7\u8bba\u70b9",
        relatedSkill: this.relatedSkill,
      });
    }

    const parallelMarkers = [
      '\u4e00\u65b9\u9762',
      '\u53e6\u4e00\u65b9\u9762',
      '\u540c\u65f6',
      '\u6b64\u5916',
    ];
    const parallelCount = parallelMarkers.reduce(
      (sum, m) => sum + content.split(m).length - 1,
      0,
    );

    if (parallelCount >= 2) {
      score += 10;
    }

    const summaryMarkers = PyramidEvaluator.LOGICAL_CONNECTORS['summary'];
    if (summaryMarkers.some((m) => content.includes(m))) {
      score += 5;
    }

    return [Math.min(100, Math.max(0, score)), issues];
  }

  private evaluateMece(
    content: string,
    paragraphs: string[],
  ): [number, Issue[]] {
    const issues: Issue[] = [];
    let score = 70;

    const frameworkSignals = [
      '\u4ece...\u89d2\u5ea6',
      '\u5206\u4e3a...\u7c7b',
      '\u5305\u62ec\u4ee5\u4e0b\u51e0\u4e2a\u65b9\u9762',
      '\u53ef\u4ee5\u5206\u4e3a',
      '\u4e3b\u8981\u6709',
      '\u4e09\u4e2a\u7ef4\u5ea6',
      '\u56db\u4e2a\u5c42\u9762',
    ];

    const hasFramework = frameworkSignals.some((s) => content.includes(s));

    if (hasFramework) {
      score += 15;
    } else {
      issues.push({
        code: 'PYRAMID_NO_CLEAR_FRAMEWORK',
        message:
          '\u672a\u4f7f\u7528\u660e\u786e\u7684\u5206\u7c7b\u6846\u67b6\uff0c\u53ef\u80fd\u5b58\u5728\u8bba\u70b9\u91cd\u53e0\u6216\u9057\u6f0f',
        severity: Severity.INFO,
        location: '\u5168\u6587',
        suggestion:
          '\u8003\u8651\u4f7f\u7528MECE\u6846\u67b6\uff08\u59823C\u30014P\u3001What/Why/How\uff09\u7ec4\u7ec7\u8bba\u70b9',
        relatedSkill: this.relatedSkill,
      });
    }

    const words = content.replace(/\n/g, ' ').split(/\s+/);
    if (words.length > 50) {
      const wordFreq: Record<string, number> = {};
      for (const word of words) {
        if (word.length > 2) {
          wordFreq[word] = (wordFreq[word] ?? 0) + 1;
        }
      }

      const highFreqWords = Object.entries(wordFreq).filter(
        ([, f]) => f > words.length * 0.05,
      );
      if (highFreqWords.length > 3) {
        score -= 10;
        issues.push({
          code: 'PYRAMID_POSSIBLE_OVERLAP',
          message:
            '\u90e8\u5206\u6982\u5ff5\u91cd\u590d\u51fa\u73b0\u9891\u7387\u8f83\u9ad8\uff0c\u53ef\u80fd\u5b58\u5728\u8bba\u70b9\u91cd\u53e0',
          severity: Severity.INFO,
          location: '\u5168\u6587',
          suggestion:
            '\u68c0\u67e5\u5404\u8bba\u70b9\u662f\u5426\u771f\u6b63\u72ec\u7acb\uff0c\u907f\u514d\u91cd\u590d\u8bba\u8ff0\u76f8\u540c\u5185\u5bb9',
          relatedSkill: this.relatedSkill,
        });
      }
    }

    return [Math.min(100, Math.max(0, score)), issues];
  }

  private splitSentences(text: string): string[] {
    return text
      .split(/[。！？.!?]/)
      .map((s) => s.trim())
      .filter((s) => s.length > 0);
  }

  private generateSummary(
    score: number,
    metrics: Record<string, number>,
    issues: Issue[],
  ): string {
    const dimensionNames: Record<string, string> = {
      conclusion_first: '\u7ed3\u8bba\u5148\u884c',
      vertical_structure: '\u7eb5\u5411\u903b\u8f91',
      horizontal_structure: '\u6a2a\u5411\u7ed3\u6784',
      mece_compliance: 'MECE\u539f\u5219',
    };

    const entries = Object.entries(metrics);
    const weakest = entries.reduce((a, b) => (b[1] < a[1] ? b : a), entries[0]);
    const strongest = entries.reduce(
      (a, b) => (b[1] > a[1] ? b : a),
      entries[0],
    );

    let summary = `\u903b\u8f91\u7ed3\u6784\u5f97\u5206\uff1a${score.toFixed(1)}/100\u3002`;
    summary += `\u6700\u5f3a\uff1a${dimensionNames[strongest[0]] ?? strongest[0]}\uff0c`;
    summary += `\u6700\u5f31\uff1a${dimensionNames[weakest[0]] ?? weakest[0]}\u3002`;

    const criticalCount = issues.filter(
      (i) => i.severity === Severity.CRITICAL,
    ).length;
    const majorCount = issues.filter(
      (i) => i.severity === Severity.MAJOR,
    ).length;

    if (criticalCount > 0) {
      summary += `\u53d1\u73b0${criticalCount}\u4e2a\u4e25\u91cd\u7ed3\u6784\u95ee\u9898\u9700\u8981\u4fee\u590d\u3002`;
    } else if (majorCount > 0) {
      summary += `\u53d1\u73b0${majorCount}\u4e2a\u4e3b\u8981\u7ed3\u6784\u95ee\u9898\u5efa\u8bae\u4f18\u5316\u3002`;
    }

    return summary;
  }

  override quickScan(content: string): EvaluationResult {
    const issues: Issue[] = [];

    const first200 =
      content.length > 200 ? content.slice(0, 200) : content;
    const hasConclusion = PyramidEvaluator.CONCLUSION_SIGNALS.some((s) =>
      first200.includes(s),
    );

    const hasStructure = PyramidEvaluator.STRUCTURE_MARKERS.some((m) =>
      content.includes(m),
    );

    let score = 50;
    if (hasConclusion) {
      score += 25;
    } else {
      issues.push({
        code: 'PYRAMID_NO_CONCLUSION_FIRST',
        message: '\u5f00\u7bc7\u53ef\u80fd\u672a\u660e\u786e\u9648\u8ff0\u7ed3\u8bba',
        severity: Severity.MAJOR,
        relatedSkill: this.relatedSkill,
      });
    }

    if (hasStructure) {
      score += 25;
    } else {
      issues.push({
        code: 'PYRAMID_NO_STRUCTURE_MARKERS',
        message: '\u7f3a\u4e4f\u7ed3\u6784\u6807\u8bb0',
        severity: Severity.MINOR,
        relatedSkill: this.relatedSkill,
      });
    }

    return new EvaluationResult(
      this.name,
      score,
      this.scoreToLevel(score),
      issues,
      {},
      `\u5feb\u901f\u626b\u63cf\uff1a\u903b\u8f91\u7ed3\u6784\u5f97\u5206${score}/100`,
    );
  }
}
